core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala

```diff
@@ -529,6 +529,46 @@ class SparkListenerSuite extends SparkFunSuite with LocalSparkContext with Match
     }
   }
 
+  Seq(true, false).foreach { throwInterruptedException =>
+    val suffix = if (throwInterruptedException) "throw interrupt" else "set Thread interrupted"
+    test(s"SPARK-30285: Fix deadlock in AsyncEventQueue.removeListenerOnError: $suffix") {
+      val conf = new SparkConf(false)
+        .set(LISTENER_BUS_EVENT_QUEUE_CAPACITY, 5)
+      val bus = new LiveListenerBus(conf)
+      val counter1 = new BasicJobCounter()
+      val counter2 = new BasicJobCounter()
+      val interruptingListener = new DelayInterruptingJobCounter(throwInterruptedException, 3)
+      bus.addToSharedQueue(counter1)
+      bus.addToSharedQueue(interruptingListener)
+      bus.addToEventLogQueue(counter2)
+      assert(bus.activeQueues() === Set(SHARED_QUEUE, EVENT_LOG_QUEUE))
+      assert(bus.findListenersByClass[BasicJobCounter]().size === 2)
+      assert(bus.findListenersByClass[DelayInterruptingJobCounter]().size === 1)
+
+      bus.start(mockSparkContext, mockMetricsSystem)
+
+      (0 until 5).foreach { jobId =>
+        bus.post(SparkListenerJobEnd(jobId, jobCompletionTime, JobSucceeded))
+      }
+
+      // Call bus.stop in a separate thread, otherwise we will block here until bus is stopped
+      val stoppingThread = new Thread(() => {
+        bus.stop()
+      })
+      stoppingThread.start()
+      // Notify interrupting listener starts to work
+      interruptingListener.sleep = false
```

Comment:
Are you trying to make sure listeners throw the exception after "stop()" is called? That's going to be hard, and your code isn't really guaranteeing that.

You could use a `CountDownLatch` that you signal right before calling `stop()` (in the thread) to unblock the listener; that will at least narrow the race down a bit.

Reply:
Maybe we could check the `stopped` status of `bus` in the listener. 
This would be better than using a `CountDownLatch`, however, it can't get rid of racing completely. WDYT?

Reply:
`CountDownLatch` always make things deterministic and it sounds better to me.

What do you mean by "it can't get rid of racing completely"?

Reply:
As the PR description, to reproduce the original issue, we have to make sure:

1.  Holding the  synchronized lock of `bus` in the stopping thread
2.  Trying to acquire the synchronized lock of `bus` in the interrupting listener thread

But signal the listener starts to interrupt just before `bus.stop` by a `CountDownLatch` can't guarantee this 100%, right?

Reply:
Maybe, you should insert `CountDownLatch` after `bus.stop`?

Reply:
I don't think so. `bus.stop` will block until all the underlying queues and listeners have been cleaned up and removed. 

> I don't think there's a way to write a proper test here without changing a bunch of things in the bus and queue code to expose internal hooks... and I don't think that's desirable.

The current test is just a trade-off. That said, it doesn't intrude into the bus code.

Reply:
An, I see, i see.

But insert `CountDownLatch` before `bus.stop` also looks good to me. Maybe, you can check `bus`'s status at the same time. e.g.

```
override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
  busStop.await(3, Seconds)
  assert(bus.isStop)
  ...
}
```

Reply:
Unfortunately,  checking the  `stoped` status can't guarantee this.  It's likely that the bus has already set the `stoped` status to true, but has not entring the synchronized the lock yet.

To avoid racing, we can pass a callback into `bus.stop`,  and notify the interrupting listener in the callback.
```scala 
  def stop(callback: Option[() => Unit]): Unit = {
    if (!started.get()) {
      throw new IllegalStateException(s"Attempted to stop bus that has not yet started!")
    }

    if (!stopped.compareAndSet(false, true)) {
      return
    }

    synchronized {
      callback.foreach{ c => c()} 
      queues.asScala.foreach(_.stop())
      queues.clear()
    }
  }

...
// in the test
bus.stop(Some(() => interruptingListener.sleep = false))

```
But that's weird. 
1.  It changes the `bus.stop` API.
2.  The synchronized in `bus.stop` has already been removed.  It's not desirable to do that for just testing an old bug.

Reply:
But your fix has removed the `synchronized`. What do you want to guarantee in the test?

Reply:
To make sure that there is no deadlock when a listener is interrupted after `bus.stop` is called. 
At the same time, we don't introduce API changing or internal status (e.g. stoped status) exposed to `bus`.

Any better idea?

Reply:
> Unfortunately, checking the stoped status can't guarantee this. It's likely that the bus has already set the stoped status to true, but has not acquired the synchronized lock yet.

IIUC, you want to let `interruptingListener` start to work once `bus` has moved to `stop` status and acquired the synchronized lock, right?

But how can `bus` acquired the synchronized lock now? This fix has already removed the synchronized lock. The only thing you could do is to check `bus` status now and I think it's enough.

Reply:
Got your point.

Now, there are two things.

1. Without the fix, what test would behave.
2. With the fix,  how to make sure that there is no deadlock when a listener is interrupted after bus.stop is called.

For (1),  we can't avoid racing without changing the `bus.stop` code (e.g. add a callback).
For (2),  we at least have to expose the internal `stoped` status of `bus`, which maybe is not recommended. 

So WDYT?

Reply:
Only focus on `LiveListenerBus` may be impossible to workaround the difficulties you mentioned above. Maybe we should move to `AsyncEventQueue`. 

How about this way:

1. Add a method `status()` in `AsyncEventQueue` for testing only;

2. In `interruptingListener`, keep checking `AsyncEventQueue.status()` until it's stopped. So, when  `AsyncEventQueue` is stopped, we're sure that `LiveListenerBus` has stopped too and acquired the lock(without fix).

WDYT?

Reply:
I believe this would work. In `AsyncEventQueue`, in fact, there is also a `stoped` status that we could check.
It sounds fine to me. cc @vanzin

Reply:
You guys are trying to fabricate a test that will not be testing what the actual code is doing when a real app is running. That's the problem.

To do that you'd need the `stop()` code in the listener bus to wait holding a lock while the queues are being drained; and one of those queues need to run into the error that causes it to remove a bad listener. That's hard to do without inserting callbacks that don't exist into the code; and adding those callbacks would only be enabling the test, which is why that's questionably.

So you basically need this in the new `stop()`:

```
def stop() {
  // do some stop stuff here
  testStartCallback()
  // clear the queues here
  testEndCallback()
}
```

The two callbacks are needed because otherwise there is no guarantee that what the queues do will happen before `stop()` does its thing.

But really I don't see what really that test would be actually testing now that there is no synchronized block anymore.

Anything you do here without these callbacks will be racy, and thus may not hit the original issue. Also, without the synchronized block, there's nothing to cause a deadlock in the first place, so that's why I said the test isn't that great to begin with.

So I'd avoid trying to create a fancy test that isn't really testing the issue and just adding unneeded hooks into the main code. The current test is ok and as close as you'll get without the above callbacks; so either go with that, or just remove the test.